lib/pbench/server/api/resources/query_apis/metadata_index/iteration_samples.py

```diff
@@ -0,0 +1,356 @@
+                  "all"
+               ],
+               "sample.measurement_title.raw":["Gb_sec"],
+               "sample.measurement_type":["throughput"],
+               "sample.name":[
+                  "sample1",
+                  "sample2",
+                  "sample3",
+                  "sample4",
+                  "sample5"
+               ],
+               "sample.server_hostname":[
+                  "127.0.0.1",
+                  "all"
+               ],
+               "sample.uid":[
+                  "client_hostname:127.0.0.1-server_hostname:127.0.0.1-server_port:20010",
+                  "client_hostname:all-server_hostname:all-server_port:all"
+               ],
+               "sample.uid_tmpl":[
+                  "client_hostname:%client_hostname%-server_hostname:%server_hostname%-server_port:%server_port%"
+               ]
+            }
+        """
+        try:
+            new_json = {
+                key: [bucket["key"] for bucket in agg["buckets"]]
+                for key, agg in es_json["aggregations"].items()
+                if agg["buckets"]
+            }
+            return jsonify(new_json)
+        except KeyError as e:
+            raise PostprocessError(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                f"Can't find Elasticsearch match data '{e}' in {es_json!r}",
+            )
+
+
+class IterationSamplesRows(RunIdBase):
+    """
+    Iteration samples API that returns iteration sample rows after
+    applying client specified filters.
+    """
+
+    DOCUMENT_SIZE = 10000  # Number of documents to return in one page
+    SCROLL_EXPIRY = "1m"  # Scroll id expires in 1 minute
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("filters", ParamType.JSON, required=False),
+                Parameter("run_id", ParamType.STRING, required=True),
+                Parameter("scroll_id", ParamType.STRING, required=False),
+            ),
+        )
+
+    def assemble(self, json_data: JSON, context: CONTEXT) -> JSON:
+        """
+        Construct a pbench Elasticsearch query for filtering iteration
+        samples based on a given run id and other filtering parameters
+        specified in JSON payload.
+
+        Note: If the ES scroll id is present we will ignore the filters
+        parameter and instead construct an Elasticsearch query for scrolling
+        based on a client provided scroll id.
+
+        If a scroll_id is specified, the query will return the next page of
+        results of the original query; otherwise we form a new query, either
+        with the specified filters or without any filters (returning up to 10000
+        documents at a time).
+
+        Args:
+            json_data:
+                "run_id": run id of a dataset that client wants to search for
+                        fetching the iteration samples.
```

Comment:
Ha. I'd commented before that `scroll_id` supersedes both `run_id` and `filters` ... you were correct in pointing out that's not true, but perhaps it's worth pointing out here in the description that (even though we won't use it in the Elasticsearch query) we *do* need `run_id` in order to authorize access to the dataset on subsequent scroll-id calls.

This may actually be an argument for moving to some internally-managed context across scrolls, because we could reliably ensure that the datasets match. (Not that it's a security hole, really, since in order to get a scroll ID you need to have supplied an initial dataset you can read; and the worst you could do by changing dataset in a follow-on scroll call would be to select a different dataset you *can't* read.)

Reply:
I concur that having that called out in a comment would be good, but I don't know that _here_ is the right place...that seems more like a schema thing.

Reply:
Actually, for just a comment, I think it would be a very good idea to include it in this PR:  it's low risk, and, if we don't do it now, when _are_ we going to do it?